### ISSUE TITLE
[FEATURE] Proposal: Introduction of @Lock Annotation for Better JDK 21 Virtual Thread Support

### ISSUE THREAD

**Issue by magicprinc** [projectlombok/lombok#3506]
Dear Project Lombok Developers,

I am writing to propose an enhancement to Project Lombok that I believe would benefit many Java developers who work with concurrent programming, particularly those using the latest Java 21 Virtual Threads.

In Java 21, the introduction of virtual threads [docs.oracle.com](https://docs.oracle.com/en/java/javase/20/core/virtual-threads.html) provides a lightweight, efficient alternative to traditional OS threads. While this is an exciting development, it presents some challenges when using the `synchronized` keyword for thread safety, especially in high load systems.

The [`java.util.concurrent.ReentrantLock` class](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/locks/ReentrantLock.html) provides an alternative to the `synchronized` keyword that offers more flexibility and can work better with virtual threads. However, using `ReentrantLock` can make the code more verbose compared to using the `synchronized` keyword.

Project Lombok's [`@Synchronized`](https://projectlombok.org/features/Synchronized) annotation has been incredibly useful in reducing boilerplate code related to synchronized methods or blocks. I would like to propose the addition of a new annotation, perhaps named `@Lock`, that would function similarly to `@Synchronized`, but use `ReentrantLock` under the hood. This would help developers leverage the benefits of `ReentrantLock` and virtual threads without adding verbosity to their code.

Here's an example of how such an annotation might be used:

```java
public class SharedObject { 
  @Lock 
  public void perform (){ 
    // Critical section - using j.u.c.Lock - here 
  } 
}
```

The `@Lock` annotation would handle the lock acquisition and release, similar to how `@Synchronized` handles synchronized blocks or methods.

Thank you for considering this proposal. I believe that the addition of this feature would greatly enhance Project Lombok and provide significant benefits to the Java developer community.

Best regards,
Andrej

**Comment by rzwitserloot** [comment]
Let's flesh it out a bit. What does `@Lock` generate if the relevant field isn't already there. How precisely does it open the lock and how precisely does it close it. Is it necessary j.u.c.ReentrantLock, or is it configurable? If configurable, in the annotation, or in lombok.config?

And, definitely, a bit of a treatise as to _why_. This is personal lack of experience. I never bother with locks, I always just use `synchronized`. If I need the power of R/W lock behaviour I'll use a juc lock (where a 'read' lock does not block out other attempts to acquire read blocks, but a 'write' block locks out all other writes _and_ reads, and open reads block out attempts to acquire writes) - but if I just need the basic lock that presumably this proposal will end up making for you, I don't get it.

Which isn't a showstopper in any way. Just a gap in my knowledge. But if I have to write the docs or at least sign off I have to understand how `synchronized` asplodes virtual threads but j.u.c. locks don't.

**Comment by Julien-Eyraud** [comment]
The main issue with `synchronized` and virtual threads is that it will pin the virtual thread to its carrier thread. To avoid this you can use another j.u.c and `ReentrantLock` seems to be a good fit for the same behavior as `syncrhonized` block.

**Comment by magicprinc** [comment]
@rzwitserloot There is a link in my post
https://docs.oracle.com/en/java/javase/20/core/virtual-threads.html
where they explain the problem with ```synchronized``` and thread "pinning"
There are also many amazing video on youtube about this subject.

TL;DR: JDK 21 Virtual Threads don’t like synchronized, and JDK dev team recommends ReentrantLock instead.

**Comment by Rawi01** [comment]
This is a duplicate of #610 and @PimvanderLoos created PR #2859 some time ago.

**Comment by rzwitserloot** [comment]
So that's heading towards the simplest possible option. No way to configure which lock, for example.

Alright. Doesn't sound particularly complicated. I need to put a little bit of thought into the name of the var we create (do we use the same name as we do for `@Synchronized`, which would imply using both in one class will clash), and more generally should we just disallow mixing `@Synchronized` and `@Lock` in the same class. I'm tempted to jus disallow that, it'd be quite crazy to want to do it.

And if we go that far... is this just a matter of adding a config key `lombok.synchronized.useJucLocks = true`? I'm pretty sure that's going way too far, and `@Synchronized` generating juc locks is a naming mismatch (a casual reader sees `@Synchronized` and assumes `synchronized (x){` instead of `new ReentrantLock`, that seems like a logical assumption, so it'd be bad if that assumption is wrong - that's why I think that's going too far).

So far I'm at:

* lombok checks if you mix `@Synchronized` and `@Lock` in one class and if so, does not make any code modification whatsoever, and generates a compiler error.
* Otherwise, lombok checks if a field named `$LOCK` / `$lock` - that's the same name `@Synchronized` uses. I have a slight preference for not clashing names, but then.. what in the blazes should we call this field if not `$lock`? (`$LOCK` for `@Lock` on static methods, `$lock` for instance methods). If this field exists, that's used. Lombok will silently assume it's properly typed and generates calls to `.lock()` and `.unlock()`; if not properly typed, lombok still does that, so you'd get a 'no such method' compiler error on the `@Lock` annotation. If the field does not exist, lombok generates it. It'll be `private final`, it'll be of type `Lock`, and its initializing expression will be `new java.util.concurrent.ReentrantLock();`
* Alternatively if the annotation is of the form `@Lock("varName")`, lombok will not generate a field in any case and proceed assuming `varName` exists and is typed `Lock` or better. Lombok doesn't check this and just generates calls to `.lock()` and `.unlock()`.
* Specifically, it generates: Add `$lock.lock();` at the top, and wrap the method content in `try { CONTENT } finally { $lock.unlock(); }`.
* `lockInterruptibly()` is not available (Unless we really wanna mess with `@Lock(interruptable = true)`, I'm not sure we should until someone explains why its important). There is no particular need to deliver on this for the first version, as we can always add it (but _removing_ it, that'd be a breaking change, hence, if we're on the fence, "do not add it yet" wins from "add it now").

Anything I've missed?

Next step is cross checking how close @Pimvanderloos 's PR is to this idea.

**Comment by magicprinc** [comment]
This is a great plan!
And PR from @PimvanderLoos is even better! And covers all my needs!

The idea about config key ```lombok.synchronized.useJucLocks = true``` sounds good at first, but is really misleading and prone to errors.

Mixing ```@Synchronized``` and ```@Lock``` in one class is possible
https://github.com/projectlombok/lombok/issues/2360
It is allowed to synchronize on instance of Lock class, but it looks more like a mistake or misunderstanding.

If someone explains cases where it can be used, you can add a config key ```lombok.synchronized.allowMixWithLock = true``` (or similar).

**Comment by rzwitserloot** [comment]
It's in :)

**Comment by magicprinc** [comment]
Sorry 🙏
Are there plans to release a new version of Lombok with this feature?

**Comment by rzwitserloot** [comment]
@magicprinc Hmm, yes, it's starting to be that kinda time. I think JDK22 GA is planned for 2024/03/19 and we certainly need a release before then, but, perhaps we shouldn't wait that long.

**Comment by rspilker** [comment]
This is part of the latest [edge release](https://projectlombok.org/download-edge), all feedback is welcome.
